iocore/net/quic/QUICStream.cc

```diff
@@ -0,0 +1,411 @@
+  this->_read_vio._cont  = nullptr;
+
+  this->_write_vio.buffer.clear();
+  this->_write_vio.nbytes = 0;
+  this->_write_vio.op     = VIO::NONE;
+  this->_write_vio._cont  = nullptr;
+}
+
+void
+QUICStream::do_io_shutdown(ShutdownHowTo_t howto)
+{
+  ink_assert(false); // unimplemented yet
+}
+
+void
+QUICStream::reenable(VIO *vio)
+{
+  if (vio->op == VIO::READ) {
+    SCOPED_MUTEX_LOCK(lock, this->_read_vio.mutex, this_ethread());
+
+    if (this->_read_vio.nbytes > 0) {
+      int event = (this->_read_vio.ntodo() == 0) ? VC_EVENT_READ_COMPLETE : VC_EVENT_READ_READY;
+
+      if (this->_read_event == nullptr) {
+        this->_read_event = this_ethread()->schedule_imm_local(this, event);
+      }
+    }
+  } else if (vio->op == VIO::WRITE) {
+    SCOPED_MUTEX_LOCK(lock, this->_write_vio.mutex, this_ethread());
+
+    if (this->_write_vio.nbytes > 0) {
+      int event = (this->_write_vio.ntodo() == 0) ? VC_EVENT_WRITE_COMPLETE : VC_EVENT_WRITE_READY;
+
+      if (this->_write_event == nullptr) {
+        this->_write_event = this_ethread()->schedule_imm_local(this, event);
+      }
+    }
+  }
+}
+
+/**
+ * @brief Signal event to this->_read_vio._cont
+ * @param (call_update)  If true, safe to call vio handler directly.
+ *   Or called from do_io_read. Still setting things up. Send event to handle this after the dust settles
+ */
+void
+QUICStream::_signal_read_event(bool direct)
+{
+  int event          = (this->_read_vio.ntodo() == 0) ? VC_EVENT_READ_COMPLETE : VC_EVENT_READ_READY;
+  Continuation *cont = this->_read_vio._cont;
+
+  if (direct) {
+    Event *e          = eventAllocator.alloc();
+    e->callback_event = event;
+    e->cookie         = this;
+    e->init(cont, 0, 0);
+
+    cont->handleEvent(event, e);
+  } else {
+    this_ethread()->schedule_imm(cont, event, this);
+  }
+}
+
+/**
+ * @brief Signal event to this->_write_vio._cont
+ * @param (call_update)  If true, safe to call vio handler directly.
+ *   Or called from do_io_write. Still setting things up. Send event to handle this after the dust settles
+ */
+void
+QUICStream::_signal_write_event(bool direct)
+{
+  int event          = (this->_write_vio.ntodo() == 0) ? VC_EVENT_WRITE_COMPLETE : VC_EVENT_WRITE_READY;
+  Continuation *cont = this->_write_vio._cont;
+
+  if (direct) {
+    Event *e          = eventAllocator.alloc();
+    e->callback_event = event;
+    e->cookie         = this;
+    e->init(cont, 0, 0);
+
+    cont->handleEvent(event, e);
+  } else {
+    this_ethread()->schedule_imm(cont, event, this);
+  }
+}
+
+void
+QUICStream::_write_to_read_vio(const std::shared_ptr<const QUICStreamFrame> &frame)
+{
+  SCOPED_MUTEX_LOCK(lock, this->_read_vio.mutex, this_ethread());
+
+  int bytes_added = this->_read_vio.buffer.writer()->write(frame->data(), frame->data_length());
+  this->_read_vio.nbytes += bytes_added;
+  this->_recv_offset += frame->data_length();
+  this->_local_flow_controller->forward_limit(frame->offset() + this->_flow_control_buffer_size);
```

Comment:
Why don't use the `this->_recv_offset` instead of `frame->offset()`. We update the offset by `frame->offset() + frame->data_length()`. we will get a negative value in forward_limit if `this->_flow_control_buffer_size` is less than `frame->data_length()`.(all this type of literal are unsigned)

Reply:
Sounds correct. Probably I forgot writing `+ frame->data_length()`.